### PR TITLE
HTCONDOR-2072 security-restruct

### PR DESCRIPTION
**Pull request by JaimeFrey**
<Insert PR description here, and leave checklist below for code review.>

# HTCondor Pull Request Checklist for internal reviewers

- [ ] Verify that (GitHub thinks) the merge is clean. If it isn't, and you're confident you can resolve the conflicts, do so. Otherwise, send it back to the original developer.
- [ ] Verify that the related Jira ticket exists and has a target version number and that it is correct.
- [ ] Verify that the Jira ticket is in review status and is assigned to the reviewer.
- [ ] Verify that the Jira ticket (HTCONDOR-xxx) is mentioned at the beginning of the title. Edit it, if not
- [ ] Verify that the branch destination of the PR matches the target version of the ticket
- [ ] Check for correctness of change
- [ ] Check for regression test(s) of new features and bugfixes (if the feature doesn't require root)
- [ ] Check for documentation, if needed  (documentation [build logs](https://readthedocs.org/projects/htcondor/builds/))
- [ ] Check for version history, if needed
- [ ] Check BaTLab dashboard for successful build (https://batlab.chtc.wisc.edu/results/workspace.php) and test for either the PR or a workspace build by the developer that has the Jira ticket as a comment.
- [ ] Check that each commit message references the Jira ticket (HTCONDOR-xxx)

## After the above
- Hit the merge button if the pull request is approved and it is not a security patch (security changes require 2 additional reviews)
- If the pull request is approved, take the ticket out of review state
- Assign JIRA Ticket back to the developer

### CODE REVIEW

**Review by GregThain — CHANGES_REQUESTED** [review]

File: src/condor_includes/KeyCache.h [hunk 1]
```diff
@@ -72,30 +67,7 @@ class KeyCacheEntry {
...
-
-	StringList * getExpiredKeys();
-
-private:
-	void copy_storage(const KeyCache &kc);
-	void delete_storage();
-
-	HashTable<std::string, KeyCacheEntry*> *key_table;
-};
+typedef std::map<std::string, KeyCacheEntry> KeyCache;
```
[the remark this answers is not in the document]
Comment (JaimeFrey) [reply src/condor_includes/KeyCache.h:70]: I'm pulling out my skinny jeans as we speak.

File: src/condor_io/condor_secman.cpp [hunk 2]
```diff
@@ -903,12 +899,9 @@ SecMan::ReconcileSecurityPolicyAds(const ClassAd &cli_ad, const ClassAd &srv_ad)
...
 
 		// send the single method for pre 6.5.0
-		StringList  tmpmethodlist( the_methods.c_str() );
-		char* first;
-		tmpmethodlist.rewind();
-		first = tmpmethodlist.next();
-		if (first) {
+		for (auto& first : StringTokenIterator(the_methods)) {
 			action_ad->Assign(ATTR_SEC_AUTHENTICATION_METHODS, first);
+			break;
```
Comment (GregThain) [inline src/condor_io/condor_secman.cpp:904]: I think we can replace this loop-with-break with

```
action_ad->Assign(ATTR_SEC_AUTHENTICATION_METHODS, *StringTokenIterator(the_methods).begin());
```

File: src/condor_io/condor_secman.cpp [hunk 3]
```diff
@@ -3292,26 +3274,19 @@ SecMan::ReconcileMethodLists( char * cli_methods, char * srv_methods ) {
...
-		if (!strcasecmp("TOKENS", sm) || !strcasecmp("IDTOKENS", sm) || !strcasecmp("IDTOKEN", sm)) {
+	for (auto& sm : StringTokenIterator(srv_methods)) {
+		if (!strcasecmp("TOKENS", sm.c_str()) || !strcasecmp("IDTOKENS", sm.c_str()) || !strcasecmp("IDTOKEN", sm.c_str())) {
 			sm = "TOKEN";
 		}
-		while ( (cm = client_methods.next()) ) {
-			if (!strcasecmp("TOKENS", cm) || !strcasecmp("IDTOKENS", cm) || !strcasecmp("IDTOKEN", cm)) {
+		for (auto cm: StringTokenIterator(cli_methods)) {
+			if (!strcasecmp("TOKENS", cm.c_str()) || !strcasecmp("IDTOKENS", cm.c_str()) || !strcasecmp("IDTOKEN", cm.c_str())) {
 				cm = "TOKEN";
```
Comment (GregThain) [inline src/condor_io/condor_secman.cpp:3287]: This isn't your change, but why are these `strcasecmp` and the above ones `strcmp` ?  If it doesn't matter, in the long run, I think I'd rather see them be strcmp, then change the literals to static std::strings which get operator== compared.

**Review by GregThain — CHANGES_REQUESTED** [review]

File: src/condor_io/condor_secman.cpp (hunk 2 above)
Comment (JaimeFrey) [reply src/condor_io/condor_secman.cpp:904]: That could crash if the list is empty. Something will have gone seriously wrong if it is empty, but I can't rule it out (especially with a malicious peer).

File: src/condor_io/condor_secman.cpp (hunk 3 above)
Comment (JaimeFrey) [reply src/condor_io/condor_secman.cpp:3287]: Those strcmp()s are the exception for methods names in the security code. All of our docs use uppercase names, but most of the code accepts both cases. If we want to change to case-sensitive comparisons, I'd want to add code to change all externally-supplied values to uppercase.

**Review by GregThain — APPROVED** [review]

**Review by GregThain — APPROVED** [review]
LGTM.

**Pull request merged**